src/node/convolver.rs

```diff
@@ -0,0 +1,568 @@
+                normalization(buffer)
+            } else {
+                1.
+            };
+
+            // Pad the response buffer with zeroes so its size is a power of 2, with 2 * 128 as min size
+            let length = buffer.length();
+            let padded_length = length.next_power_of_two().max(2 * RENDER_QUANTUM_SIZE);
+            let samples: Vec<_> = (0..buffer.number_of_channels())
+                .map(|_| {
+                    let mut samples = vec![0.; padded_length];
+                    samples[..length]
+                        .iter_mut()
+                        .zip(buffer.get_channel_data(0))
+                        .for_each(|(o, i)| *o = *i * scale);
+                    samples
+                })
+                .collect();
+
+            let padded_buffer = AudioBuffer::from(samples, sample_rate);
+
+            let convolve = ConvolverRendererInner::new(padded_buffer);
+            let _ = self.sender.send(convolve); // can fail when render thread shut down
+        }
+
+        *self.buffer.lock().unwrap() = buffer;
+    }
+
+    /// Denotes if the response buffer will be scaled with an equal-power normalization
+    pub fn normalize(&self) -> bool {
+        self.normalize
+    }
+
+    /// Update the `normalize` setting. This will only have an effect when `set_buffer` is called.
+    pub fn set_normalize(&mut self, value: bool) {
+        // TODO, use AtomicBool to prevent &mut self?
+        self.normalize = value;
+    }
+}
+
+fn roll_zero<T: Default + Copy>(signal: &mut [T], n: usize) {
+    // roll array by n elements
+    // zero out the last n elements
+    let len = signal.len();
+    signal.copy_within(n.., 0);
+    signal[len - n..].fill(T::default());
+}
+
+struct Fft {
+    fft_planner: RealFftPlanner<f32>,
+    fft_input: Vec<f32>,
+    fft_scratch: Vec<Complex<f32>>,
+    fft_output: Vec<Complex<f32>>,
+}
+
+impl Fft {
+    fn new(length: usize) -> Self {
+        let mut fft_planner = RealFftPlanner::<f32>::new();
+        let fft = fft_planner.plan_fft_forward(length);
+        let fft_input = fft.make_input_vec();
+        let fft_scratch = fft.make_scratch_vec();
+        let fft_output = fft.make_output_vec();
+
+        Self {
+            fft_planner,
+            fft_input,
+            fft_scratch,
+            fft_output,
+        }
+    }
+
+    fn real(&mut self) -> &mut [f32] {
+        &mut self.fft_input[..]
+    }
+
+    fn complex(&mut self) -> &mut [Complex<f32>] {
+        &mut self.fft_output[..]
+    }
+
+    fn process(&mut self) -> &[Complex<f32>] {
+        let fft = self.fft_planner.plan_fft_forward(self.fft_input.len());
+        fft.process_with_scratch(
+            &mut self.fft_input,
+            &mut self.fft_output,
+            &mut self.fft_scratch,
+        )
+        .unwrap();
+        &self.fft_output[..]
+    }
+
+    fn inverse(&mut self) -> &[f32] {
+        let fft = self.fft_planner.plan_fft_inverse(self.fft_input.len());
```

Comment:
maybe `fft` and `fft_inverse` could be created once for all in `new`, I saw `fft_planner` has some cache system but it looks a bit like unnecessary work

Reply:
Yes right, we could store the `Arc<dyn FFT>`s directly instead of the planner. Will do that